text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,455 @@
+          "sha256": "node14.5.0-sha256",
+          "id": "node",
+          "version": "14.5.0",
+          "stacks": [
+            "io.buildpacks.stacks.bionic"
+          ]
+        }
+      ]
+    }
+  }
+}
+```
+
+
+This label let us determine:
+- set of all assets that could be used by a buildpackage
+- set of assets that may be used by an individual buildpack.
+- locations to fetch assets from
+
+
+To keep our metadata consistent we also propose adding the `assets` array to the `io.buildpacks.buildpackage.metadata` label.
+
+##### General Format
+```json
+{
+  "id": "buildpack-id",
+  "version": "0.0.1",
+  "homepage": "https:/buildpacks.io",
+  "stacks": [
+    {
+      "id": "io.buildpacks.stacks.bionic"
+    }
+  ],
+  "assets": [
+    {
+      "sha256": "asset-1-sha256",
+      "id": "asset-1-id",
+      "version": "1.2.3",
+      "stacks": [
+        "stack1",
+        "..."
+      ]
+    },
+    {
+      "sha256": "asset-2-sha256",
+      "id": "asset-2-id",
+      "version": "2.3.4",
+      "stacks": [
+        "stack1",
+        "..."
+      ]
+    },
+    "..."
+  ]
+}
+```
+
+
+### Builder Changes
+
+To keep track of what assets are actually in our builder, and what layer they are contained in we add the label, `io.buildpacks.assets.layers`. Note, this label has already been defined above on asset-cache.
+
+Builders will also require the same changes outlined in the Buildpackage section above to the `io.buildpacks.buildpack.layers` tag.
+
+Additionally we will add an `asset` array to `io.buildpacks.builder.metadata` to keep it consistent with the updates to  `io.buildpacks.builpack.layers`.
+
+##### General Format
+```json
+{
+  "buildpacks": [
+    {
+      "id": "buildpack-id",
+      "version": "0.0.1",
+      "homepage": "https://buildpacks.io",
+      "assets": [
+        {
+          "sha256": "asset-1-sha256",
+          "id": "asset-1-id",
+          "version": "1.2.3",
+          "stacks": [
+            "stack1",
+            "..."
+          ]
+        },
+        {
+          "sha256": "asset-2-sha256",
+          "id": "asset-2-id",
+          "version": "2.3.4",
+          "stacks": [
+            "stack1",
+            "..."
+          ]
+        },
+        "..."
+      ]
+    }
+  ]
+}
+```
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Asset cache creation
+
+Asset Image creation will be handled by the platform. E.g. `pack package-asset <asset-cache-name> --asset-config <optional-path-to-asset.toml>`.
+
+There are two ways to package assets
+1) from a `buildpack.toml` file. This method uses the `assets` array to locate download and verify assets.
+    - Ex: `pack package-asset <asset-cache-name> <path-to-buildpack.toml>`
+2) from an already existing buildpackage. This will download all assets using `uri`s from the `io.buildpacks.buildpack.layers` label, or re-use assets that are included in the specified buildpackage's layers.
+    - Ex: `pack package-asset <asset-image-name> <path-to-package.toml>`
```

Comment:
If we want to package from an already existing buildpackage, it might make more sense to:
`pack package-asset <asset-image-name> <buildpack-uri>`
As the buildpackage itself is the artifact that is passed around (and stored in the registry) rather than `package.toml`.

What is nice about this is that since a directory containing a `buildpack.toml` is also a buildpack URI (as pack understands it) so this usage could also satisfy the above case.